### PR TITLE
Spatial smoothing algorithm in Dynsight

### PR DESCRIPTION
**Pull request by SimoneMartino98**
With this pull request I want to add the spatial smoothing parallelized algorithm.

In brief, this code can be useful when the user wants to average the value of a collective variable of a particle with the values of its neighboring particles, defined by a cutoff distance.

### CODE REVIEW

File: src/dynsight/_internal/data_processing/spatial_smoothing.py [hunk 1]
```diff
@@ -62,9 +64,10 @@ def spatial_smoothing(
 
     shape = array.shape
     dtype = array.dtype
-
     shared_array = Array(ctypes.c_double, array.size, lock=False)
-    shared_array_np = np.frombuffer(shared_array, dtype=dtype).reshape(shape)
+    # The mypy error [call-overload] is ignored as it is considered.
```
Comment (andrewtarzia) [inline src/dynsight/_internal/data_processing/spatial_smoothing.py:68]: Remove the full stop at the end.

Also, a bit more detail is needed: `np.frombuffer type is 'buffer-like', which contains 'Array'`

**Review by andrewtarzia — CHANGES_REQUESTED** [review]

File: docs/source/data_processing.rst [hunk 2]
```diff
@@ -29,3 +29,11 @@ Classification
   savereferences <_autosummary/dynsight.data_processing.savereferences>
   getreferencesfromdataset <_autosummary/dynsight.data_processing.getreferencesfromdataset>
   applyclassification <_autosummary/dynsight.data_processing.applyclassification>
+
+Pre-analysis tools
```
[the remark this answers is not in the document]
Comment (SimoneMartino98) [reply docs/source/data_processing.rst]: done

File: src/dynsight/_internal/data_processing/spatial_average.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    # of the code. Ruff error PLW0603 is therefore ignored.
+    global array  # noqa: PLW0603
+    array = np.frombuffer(shared_array, dtype=dtype).reshape(shape)
+
+
+def processframe(args: Any) -> tuple[int, np.ndarray[float, Any]]:
+    universe, selection, cutoff, frame, vector = args
+    universe.trajectory[frame]
+    distances = distance_array(
+        selection.positions, selection.positions, box=universe.dimensions
```
[the remark this answers is not in the document]
Comment (SimoneMartino98) [reply src/dynsight/_internal/data_processing/spatial_average.py:31]: done

File: src/dynsight/_internal/data_processing/spatial_average.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    Example:
+
+        .. code-block:: python
+
+            from dynsight.data_processing import spatialaverage
+
+            u = MDAnalysis.Universe('topology.gro', 'trajectory.xtc')
+            averaged_values = spatialaverage(
+                                universe = u,
```
[the remark this answers is not in the document]
Comment (SimoneMartino98) [reply src/dynsight/_internal/data_processing/spatial_average.py:123]: only in the docs or in the entire code ?

[the remark this answers is not in the document]
Comment (andrewtarzia) [reply src/dynsight/_internal/data_processing/spatial_average.py]: Only when it is for a keyword argument, like in this line of code

[the remark this answers is not in the document]
Comment (SimoneMartino98) [reply src/dynsight/_internal/data_processing/spatial_average.py]: ok; I'll do it

File: src/dynsight/_internal/data_processing/spatial_average.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+from pathlib import Path
+
+import numpy as np
+from MDAnalysis.analysis.distances import distance_array
+
+array: np.ndarray[float, Any]
+
+
+def initworker(
+    shared_array: np.ndarray[float, Any], shape: int, dtype: tuple[float, Any]
```
Comment (andrewtarzia) [inline src/dynsight/_internal/data_processing/spatial_average.py:19]: bring `shape` to the next line.

File: src/dynsight/_internal/data_processing/spatial_average.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            defines the neighborhood radius within which particles are
+            considered as neighbors.
+        traj_cut:
+            Number of frames to exclude from the end of the trajectory.
+        num_processes:
+            The number of processes to use for parallel computation.
+            **Warning** adjust this based on the available cores.
+
+    Returns:
+        np.ndarray
```
Comment (andrewtarzia) [inline src/dynsight/_internal/data_processing/spatial_average.py:105]: Remove this typing.

**Comment by andrewtarzia** [comment]
Almost there, a few unresolved conversations above.

File: src/dynsight/_internal/data_processing/spatial_average.py [hunk 7]
```diff
@@ -84,7 +83,7 @@ def spatialaverage(
         universe:
             The MDAnalysis `Universe` object containing the molecular dynamics
             simulation data, including atom positions and trajectory.
-        array_path:
+        descriptor_array:
             Path to the file containing the NumPy array of descriptor values.
```
Comment (andrewtarzia) [inline src/dynsight/_internal/data_processing/spatial_average.py:87]: Minor fix.

**Review by andrewtarzia — APPROVED** [review]

**Pull request merged**